docs/RFCS/20171004_sql_sequences.md

```diff
@@ -0,0 +1,427 @@
+      The introspectability could be achieved by creating a virtual table.
+      (The `information_schema.sequences` table, which we'll be implementing,
+      gets you some of this: all the metadata, just not the value.)
+
+### Impact of not doing this
+
+Users who prefer sequences or who have applications which use them will have
+an obstacle in moving to Cockroach.
+
+## Unresolved questions
+
+### Design / scope issues
+
+- Should we change our `SERIAL` type to be backed by sequences? I don't think
+  we can do this until we make it faster by implementing `CACHE` (see below).
+
+  We would also have to decide whether to make `SERIAL` int32. While more
+  compatible with Postgres, this would be backwards-incompatible with Cockroach,
+  since our current `SERIAL` is int64.
+
+  Making `SERIAL` always sequence backed would also allow us to automatically
+  create a sequence when a `SERIAL` is used in a `CREATE TABLE`, as Postgres
+  does. With the implementation in this RFC, to use sequences users will have to
+  explicitly type `CREATE SEQUENCE` and set their column to
+  `DEFAULT nextval('my_sequence')`.
+- Should we implement the `AS` clause, which enables users to specify which
+  type they want the sequence value to be? I think not, since it works to
+  create a `SMALLINT` column that uses `DEFAULT nextval(seq)` and get a value,
+  even though `nextval` returns `int64` and `SMALLINT` is `int32`. Silent
+  truncation should not occur here, since we [check the size of the value before
+  inserting it][value-size-check-gh] and return an error.
+- Should we implement the `CACHE` setting? (pre-allocation on each node of
+  batches of values, which can then be handed out quickly from memory)
+  This would probably make sequences fast enough to be our default for the
+  `SERIAL` type, but I think we should build the simple version, benchmark,
+  and do `CACHE` in a later RFC/PR.
+- Should we make `SELECT * FROM my_sequence` work? In Postgres, it returns a
+  single row containing the sequence value and settings:
+
+  | sequence_name | last_value | start_value | increment_by |      max_value      | min_value | cache_value | log_cnt | is_cycled | is_called |
+  |---------------|------------|-------------|--------------|---------------------|-----------|-------------|---------|-----------|-----------|
+  | foo           |          2 |           1 |            1 | 9223372036854775807 |         1 |           1 |      31 | f         | t         |
+
+  This is easy to do, and may be helpful for compatibility. It can be saved for
+  the last PR of the change, but should probably be done.
+- Should the sequence values be exposed as a `system.sequences` table, for easy
+  introspectability? Ideally yes, if we can still use the KV-level `Inc`
+  operation. I will try this during implementation.
+
+### Implementation issues
+
+Currently, some functions are nondeterministic (e.g. `random`), but none write
```

Comment:
@knz, any thoughts on this? It would seem to involve adding some kind of KV-layer access to the `EvalPlanner` that gets passed to functions.